### PR TITLE
Add Hylo-demangle tool

### PR DESCRIPTION
**Pull request by lucteo**
The tool parses the input and replaces the symbols that look like Hylo mangled symbols with their demangled string, keeping everything else the same.

Add "$h" as a prefix for Hylo mangled names.

### CODE REVIEW

File: Sources/hylo-demangle/CommandLine.swift [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+  /// Executes the command.
+  public func run() throws {
+    let input = try sourceText()
+
+    if listOnly {
+      input.enumerateTokens { (token) in
+        if let demangled = String(token).demangled() {
+          print("\(token) => \(demangled)")
+        }
+      }
```
Comment (tothambrus11) [inline Sources/hylo-demangle/CommandLine.swift]: This could be factored out to a function, then we can avoid the early return and just have a symmetrical-looking `if`/`else`.

**Review by tothambrus11 — CHANGES_REQUESTED** [review]
Nice work! 
If we are fine with keeping the whole processed text in memory, we could refactor the printing to accumulate into a string, so that the code becomes easier to test (the corner cases of the index handling and making sure nothing from the content gets dropped).

File: Tests/BackEndTests/SimpleFunctionEmitterTest.swift [hunk 2]
```diff
@@ -31,7 +31,7 @@ final class SimpleFunctionEmitterTest: XCTestCase {
     XCTAssert(
       llvmIR.containsOnce(
         """
-        define private void @"M6testvUpvirtualu6rky9fkv9x3F06mainlT01tR50tR5$gP71L1L1L"() {
```
Comment (tothambrus11) [inline Tests/BackEndTests/SimpleFunctionEmitterTest.swift:34]: It's really nice that most of the mangling tests could remain the same, while we still got this indicative change here :D

File: Package.swift [hunk 3]
```diff
@@ -59,6 +60,14 @@ let package = Package(
       ],
       swiftSettings: commonSwiftSettings),
 
+    .executableTarget(
+      name: "hylo-demangle",
```
Comment (tothambrus11) [inline Package.swift:64]: The current pattern seems to be to prefix CLI tools with `hc-` (e.g., `hc-tests`).

File: Sources/FrontEnd/IR/Mangling/DemangledSymbol.swift [hunk 4]
```diff
@@ -5,7 +5,13 @@ internal indirect enum DemangledSymbol: Hashable, Sendable {
   ///
   /// `self` is assigned to an error if `s` is malformed.
```
Comment (tothambrus11) [inline Sources/FrontEnd/IR/Mangling/DemangledSymbol.swift]: I'm wondering if we should describe `.notMangled` here.

File: Package.swift (hunk 3 above)
Comment (lucteo) [reply Package.swift:64]: I tried to keep the name we had in the old compiler. I feel that using `hc-` prefix is not thee most appropriate, as this is not a direct part of the compiler

Comment (tothambrus11) [reply Package.swift:64]: I think I also like the `hylo-` prefix more, it's also less obfuscated.

File: Sources/hylo-demangle/CommandLine.swift [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+        print(token, terminator: "")
+      }
+      lastIndex = token.endIndex
+    }
+    if lastIndex < input.endIndex {
+      print(input[lastIndex..<input.endIndex], terminator: "")
+    }
+  }
+
+  /// Prints the demangled form of each matched mangled symbol om `input`.
```
Comment (tothambrus11) [inline Sources/hylo-demangle/CommandLine.swift]:
```suggestion
  /// Prints the demangled form of each matched mangled symbol in `input`.
```

File: Sources/hylo-demangle/CommandLine.swift [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+        print(demangled, terminator: "")
+      } else {
+        print(token, terminator: "")
+      }
+      lastIndex = token.endIndex
+    }
+    if lastIndex < input.endIndex {
+      print(input[lastIndex..<input.endIndex], terminator: "")
+    }
+  }
```
Comment (tothambrus11) [inline Sources/hylo-demangle/CommandLine.swift:68]: It would be nice to have some tests for this code, especially its edge cases. We could accumulate into a string and print only at the call site.

Comment (tothambrus11) [reply Sources/hylo-demangle/CommandLine.swift:68]: We decided that it wasn't worth the refactor for now.

**Review by tothambrus11 — APPROVED** [review]
